File: public/dsa/linked_lists/find_xth_node_from_end_of_linked_list/approach/find_xth_node_from_end_of_linked_list.py
# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def xthNodeFromEnd(self, head: ListNode, x: int) -> ListNode:
        fast = head
        for _ in range(x):
            if not fast:
                return None  # x > length
            fast = fast.next

        slow = head
        while fast:
            fast = fast.next
            slow = slow.next

        return slow


# tests-start
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
sol = Solution()
print(sol.xthNodeFromEnd(head, 2).data)
# tests-end

'''output
3
'''

File: public/dsa/linked_lists/find_xth_node_from_end_of_linked_list/approach/find_xth_node_from_end_of_linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

// defs-start
struct ListNode {
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};
// defs-end

ListNode* xthNodeFromEnd(ListNode* head, int x) {
    if (!head) return NULL;

    ListNode* fast = head;
    ListNode* slow = head;

    for (int i = 0; i < x; i++) {
        if (!fast) return NULL;
        fast = fast->next;
    }

    while (fast) {
        fast = fast->next;
        slow = slow->next;
    }

    return slow;
}

// tests-start
int main() {
    ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));

    cout << xthNodeFromEnd(head, 2)->data << endl;
    return 0;
}
// tests-end

/*output
3
*/
